### ISSUE TITLE
connect EMFILE

### ISSUE THREAD

**Issue by rooi** [KhaosT/HAP-NodeJS#53]
Hi KhaosT,

I recently get errors like the ones below and HAP-NodeJS stops working after a while (say 1 day). Any idea what could cause them?

I using the BridgedCore.js from HAP-NodeJS on a mac and an iPhone 4s on iOS 8.3.
I've updated the mdns module upon your earlier suggestion to fix bonjour problems

When I lookup the error myself it seems like there are too many open connections. Could it be them some connections are not closed properly? Could you give me some pointers to debug this problem?

Thanks!

Roy

...
...
...
Start Verify M1
Verify M1 Finished
Start Verify M3
M3: Verify Success
New Session 53171
Server Connection Established 63241
Start Verify M1
Verify M1 Finished
An Error Occured on client side connection, { [Error: read ECONNRESET] code: 'ECONNRESET', errno: 'ECONNRESET', syscall: 'read' }
Server Disconnected
Client Disconnected
New Session 53173
Server Connection Established 63247
...
...
...
Start Verify M3
M3: Verify Success
New Session 53284
Server Connection Established 57015
Start Verify M1
Verify M1 Finished
Start Verify M3
M3: Verify Success
New Session 53287
An Error Occured on HAP side connection, { [Error: connect EMFILE] code: 'EMFILE', errno: 'EMFILE', syscall: 'connect' }
Client Disconnected
Server Disconnected
New Session 53289
An Error Occured on HAP side connection, { [Error: connect EMFILE] code: 'EMFILE', errno: 'EMFILE', syscall: 'connect' }
Client Disconnected
Server Disconnected
...
...

**Comment by KhaosT** [comment]
@rooi Yeah, there might be some issue with socket managements. I'll try to fix that when I have time, meanwhile, you can follow [this](http://stackoverflow.com/questions/10355501/connect-emfile-error-in-node-js) and temporarily improve the max files limit in linux

**Comment by rooi** [comment]
Thanks @KhaosT. If there is something I can do to help...

**Comment by KhaosT** [comment]
I believe the refactor version of HAP-NodeJS fixed the issue.

**Comment by desvdp** [comment]
Some observations:
- I'm running on an RPi, which had the ulimit for open files set to 1024: I tend to believe this is just not enough for running the system. On a clear day, with HAP-Nodejs lsof | grep "" -c reports over 1024 open "files". So I increased the number to start with... so far no big deal.
- When counting the TCP sockets that are open (lsof | grep IP | grep the pid of the process), I get "50". Two things come up in my mind: given I'm running in bridged mode I figured this was quite a lot, but then again I do have 40 accessories behind the bridge. So I guess even in bridged mode each accessory gets a socket, is this correct ?
- One thing I did note: when I disable my iPhone Wifi it does take a couple of minutes to close the stale sockets, when reopening the connection 50 goes 90 ... and later back to 50.  

Bottom line: I tend to believe the socket management works fine, but in contradiction to the article above increasing the ulimit is probably a good idea after all... (mine is set to 4096 now).

**Comment by KhaosT** [comment]
@desvdp are you sure you're in bridged mode? if so, there might have something wrong... it should only have one connection to your iOS device and one connection to local HTTP server... 50 is definitely wrong...

**Comment by desvdp** [comment]
@KhaosT I'd say I am: when looking for accessories on my iPhone I only see the bridge device (all the others come in nicely with that one). I must say I am running via homebridge shims, although I don't believe that has anything to doe with it. Regarding homebridge: I am running two platform shims, I have 6 Hue lamps, plus 33 devices on my own custom platform (Nikobus) plus the bridge  which adds up to the 40 I mentioned above... 
On the iOS side of things I'm on the latest 8, no iOS 9 beta...

**Comment by KhaosT** [comment]
@desvdp and how many iOS devices are connected to the server? Can you post the result of ``lsof -p ServerPID | grep IP`` at here?

**Comment by desvdp** [comment]
@khaost Just one iOS device ... I'm happy to post the lsof result, but I'm affraid I'm currently at the office. I'll see what I can do to get them to you asap...

**Comment by desvdp** [comment]
@khaost when my iOS device is not connected I get only 1 socket if I use the -p option you specified. When I do a grep on the PID as I did I get 10 (which is expected by as 50 - 40), I guess these are just listeners in sub-processes. So I assume you are correct: only one socket is open.  But the question is which is the relevant one for the ulimit count? 

> pi@raspberrypi ~ $ lsof -p 23541 | grep IP
> node    23541   pi   14u  IPv4      42645      0t0    TCP *:51826 (LISTEN)
>
> pi@raspberrypi ~ $ lsof  | grep IP | grep 23541
> node      23541               pi   14u     IPv4      42645      0t0    TCP *:51826 (LISTEN)
> Optimizin 23541 23542         pi   14u     IPv4      42645      0t0    TCP *:51826 (LISTEN)
> v8:Sweepe 23541 23543         pi   14u     IPv4      42645      0t0    TCP *:51826 (LISTEN)
> v8:Sweepe 23541 23544         pi   14u     IPv4      42645      0t0    TCP *:51826 (LISTEN)
> v8:Sweepe 23541 23545         pi   14u     IPv4      42645      0t0    TCP *:51826 (LISTEN)
> node      23541 23546         pi   14u     IPv4      42645      0t0    TCP *:51826 (LISTEN)
> node      23541 23547         pi   14u     IPv4      42645      0t0    TCP *:51826 (LISTEN)
> node      23541 23548         pi   14u     IPv4      42645      0t0    TCP *:51826 (LISTEN)
> node      23541 23549         pi   14u     IPv4      42645      0t0    TCP *:51826 (LISTEN)
> node      23541 23552         pi   14u     IPv4      42645      0t0    TCP *:51826 (LISTEN)

**Comment by KhaosT** [comment]
@desvdp not sure about that, you'll have to ask Google :)

**Comment by desvdp** [comment]
@KhaosT  Tried that earlier ... Google didn't seem to agree on it ;-) . 
In any case: if the setup stops working, in my opinion it's worth having a look at the ulimit...
